pkg/assets/create/creater.go

```diff
@@ -0,0 +1,190 @@
+	// Filters allows to filter which files we will read from disk.
+	Filters []assets.FileInfoPredicate
+}
+
+// EnsureManifestsCreated ensures that all resource manifests from the specified directory are created.
+// This function will keep retrying creation until no errors are reported.
+// Pass the context to indicate how much time you are willing to wait until all resources are created.
+func EnsureManifestsCreated(ctx context.Context, manifestDir string, restConfig *rest.Config, options CreateOptions) error {
+	client, err := dynamic.NewForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	dc, err := discovery.NewDiscoveryClientForConfig(restConfig)
+	if err != nil {
+		return err
+	}
+
+	manifests, err := load(manifestDir, options)
+	if err != nil {
+		return err
+	}
+
+	getRESTMapper := func() (meta.RESTMapper, error) {
+		gr, err := restmapper.GetAPIGroupResources(dc)
+		if err != nil {
+			return nil, err
+		}
+		return restmapper.NewDiscoveryRESTMapper(gr), nil
+	}
+
+	mapper, err := getRESTMapper()
+	if err != nil {
+		return err
+	}
+
+	// Retry creation until no errors are returned or the timeout is hit.
+	var lastCreateError error
+	err = wait.PollImmediateUntil(500*time.Millisecond, func() (bool, error) {
+		err, refresh := create(manifests, client, mapper, options)
+		if err == nil {
+			// No errors means all resources were created successfully, reset lastCreateError and exit the loop.
+			lastCreateError = nil
+			return true, nil
+		}
+		// We got rest-mapper error, force to refresh discovery so next create use updated mapper.
+		if refresh {
+			mapper, err = getRESTMapper()
+			if err != nil {
+				return false, err
+			}
+		}
+		lastCreateError = err
+		return false, nil
+	}, ctx.Done())
+
+	// Return the last observed set of errors from the create process instead of timeout error.
+	if lastCreateError != nil {
+		return lastCreateError
+	}
+
+	return err
+}
+
+func create(manifests map[string]*unstructured.Unstructured, client dynamic.Interface, mapper meta.RESTMapper, options CreateOptions) (error, bool) {
+	// Sort all manifests, so in case they use number prefixes, we follow their order, which will increase
+	// the chances for the create loop to succeed faster.
+	sortedManifestPaths := []string{}
+	for key := range manifests {
+		sortedManifestPaths = append(sortedManifestPaths, key)
+	}
+	sort.Strings(sortedManifestPaths)
+
+	// Record all errors for the given manifest path (so when we report errors, users can see what manifest failed).
+	errs := map[string][]error{}
+
+	// In case we fail to find a rest-mapping for the resource, force to fetch the updated discovery on next run.
+	reloadDiscovery := false
+
+	for _, path := range sortedManifestPaths {
+		gvk := manifests[path].GetObjectKind().GroupVersionKind()
+		mappings, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+		if err != nil {
+			errs[path] = append(errs[path], fmt.Errorf("unable to get REST mapping: %v", err))
+			reloadDiscovery = true
```

Comment:
And `continue`?

Reply:
missing continue here is a logical bug, fixed

Reply:
i re-use formatErrors in "load"

Reply:
I've pushed https://github.com/openshift/library-go/compare/master...wking:format-errors-drop-slice a4bb2915a9, dropping the errors slice, to address the "never have more than one error per path" portion of this thread.

Reply:
> I've pushed https://github.com/wking/openshift-library-go/tree/format-errors-drop-slice a4bb291...

Oops, missed one line here.  Fixed with a4bb291 -> c009e95.

Reply:
thanks! fixed up to the PR.